File: user/src/lib/netstats.zig
//! Userland mirror of the kernel's `sys_net_stats` snapshot (slot 62).
//!
//! The struct below is a byte-for-byte mirror of `kernel/src/syscall.zig`'s
//! `NetStats` extern struct. Both sides pin the key `@offsetOf`s and the
//! struct size in host tests, so any drift fails loudly on the next test
//! run — the kernel test "sys_net_stats marshals a whole snapshot and
//! pins the layout" plus the size test at the bottom of this file.
//!
//! All integers little-endian; IPs are raw network-order bytes ([4]u8);
//! MACs raw bytes ([6]u8). Enum naturals are pinned in the kernel doc.

const std = @import("std");
const ui = @import("ui.zig");

pub const NetStats = extern struct {
    // ---- interface ---------------------------------------------------------
    mac: [6]u8 = .{ 0, 0, 0, 0, 0, 0 },
    own_ip: [4]u8 = .{ 0, 0, 0, 0 },
    gateway: [4]u8 = .{ 0, 0, 0, 0 },
    // ---- dhcp ---------------------------------------------------------------
    dhcp_state: u8 = 0, // 0 idle,1 selecting,2 requesting,3 bound,4 renewing,5 rebinding
    lease_ip: [4]u8 = .{ 0, 0, 0, 0 },
    lease_mask: [4]u8 = .{ 0, 0, 0, 0 },
    lease_server: [4]u8 = .{ 0, 0, 0, 0 },
    lease_secs: u32 = 0,
    // ---- TCP ----------------------------------------------------------------
    tcp_state: u8 = 0, // 0 idle,1 syn_sent,2 established,3 fin_sent,4 closed
    tcp_peer_ip: [4]u8 = .{ 0, 0, 0, 0 },
    tcp_peer_port: u16 = 0,
    // ---- UDP ----------------------------------------------------------------
    udp_count: u8 = 0,
    udp_ports: [4]u16 = .{ 0, 0, 0, 0 },
    // ---- ARP ----------------------------------------------------------------
    arp_count: u8 = 0,
    arp_ips: [4][4]u8 = .{ .{ 0, 0, 0, 0 }, .{ 0, 0, 0, 0 }, .{ 0, 0, 0, 0 }, .{ 0, 0, 0, 0 } },
    arp_macs: [4][6]u8 = .{ .{ 0, 0, 0, 0, 0, 0 }, .{ 0, 0, 0, 0, 0, 0 }, .{ 0, 0, 0, 0, 0, 0 }, .{ 0, 0, 0, 0, 0, 0 } },
    // ---- counters -----------------------------------------------------------
    tx_frames: u64 = 0,
    tx_bytes: u64 = 0,
    rx_frames: u64 = 0,
    rx_bytes: u64 = 0,
    rx_filtered: u64 = 0,
    rx_overflow: u64 = 0,
    // tcp segment counters: syn_sent, synack_recv, ack_sent, data_sent,
    // data_recv, fin_sent, finack_recv, rst_sent
    tcp_segs: [8]u64 = .{ 0, 0, 0, 0, 0, 0, 0, 0 },
    // udp datagram counters: received, sent, loopbacked, dropped
    udp_dgrams: [4]u64 = .{ 0, 0, 0, 0 },
};

pub const net_stats_bytes: usize = @sizeOf(NetStats);

/// Fetch one snapshot. Returns true on success (snapshot fully copied);
/// false when the kernel refused (too-small buffer / EFAULT — the caller
/// keeps its previous view).
pub fn read_stats(snap: *NetStats) bool {
    const rc = ui.syscall2(ui.sys_net_stats_num, @intFromPtr(snap), net_stats_bytes);
    if (rc != net_stats_bytes) return false;
    return true;
}

// Human-readable helpers (pure — host-tested).

pub fn dhcp_state_name(state: u8) []const u8 {
    return switch (state) {
        1 => "selecting",
        2 => "requesting",
        3 => "bound",
        4 => "renewing",
        5 => "rebinding",
        else => "idle",
    };
}

pub fn tcp_state_name(state: u8) []const u8 {
    return switch (state) {
        1 => "SYN-SENT",
        2 => "ESTABLISHED",
        3 => "FIN-WAIT",
        4 => "CLOSED",
        else => "IDLE",
    };
}

test "netstats: struct size and key offsets match the kernel pins" {
    try std.testing.expectEqual(@as(usize, 6), @offsetOf(NetStats, "own_ip"));
    try std.testing.expectEqual(@as(usize, 10), @offsetOf(NetStats, "gateway"));
    try std.testing.expectEqual(@as(usize, 14), @offsetOf(NetStats, "dhcp_state"));
    try std.testing.expectEqual(@as(usize, 28), @offsetOf(NetStats, "lease_secs"));
    try std.testing.expectEqual(@as(usize, 32), @offsetOf(NetStats, "tcp_state"));
    try std.testing.expectEqual(@as(usize, 33), @offsetOf(NetStats, "tcp_peer_ip"));
    try std.testing.expectEqual(@as(usize, 38), @offsetOf(NetStats, "tcp_peer_port"));
    try std.testing.expectEqual(@as(usize, 40), @offsetOf(NetStats, "udp_count"));
    try std.testing.expect(net_stats_bytes <= 256);
}

test "netstats: state name helpers" {
    try std.testing.expectEqualStrings("bound", dhcp_state_name(3));
    try std.testing.expectEqualStrings("idle", dhcp_state_name(0));
    try std.testing.expectEqualStrings("ESTABLISHED", tcp_state_name(2));
    try std.testing.expectEqualStrings("CLOSED", tcp_state_name(4));
}

File: user/go/vi/netstats.go
package vi

// M71g (issue #1566): the Go mirror of `sys_net_stats` (ADR 0007 slot 62).
//
// The kernel owns the wire format (kernel/src/syscall.zig `NetStats`,
// marshalled by `handle_net_stats`); the userland mirror lives in
// user/src/lib/netstats.zig. This struct is the Go twin of that mirror, and
// netstats_test.go pins the SAME key offsets and the same 240-byte size, so
// drift on any side fails a test rather than a dashboard reading silently.
//
// Semantics of the call itself, per the kernel handler:
//   - `max < net_stats_bytes` returns 0 — a REFUSAL, not an error (nothing is
//     copied; the caller keeps its previous view).
//   - otherwise it copies the whole snapshot out and returns its byte size.
//   - a bad destination pointer is -EFAULT.
//   - All integers little-endian; IPs are raw network-order bytes ([4]byte);
//     MACs raw bytes ([6]byte). Enum naturals are pinned in the kernel doc.
//
// Off the guest the call degrades to -ENOSYS (vi_host.go), so the decode path
// is exercised through the syscallHook seam.

import "unsafe"

// NetStatsBytes is the pinned snapshot size (`@sizeOf(NetStats)` on both
// other sides; 240 with the padding the layout forces).
const NetStatsBytes = 240

// NetStats is one decoded snapshot.
type NetStats struct {
	MAC       [6]byte
	OwnIP     [4]byte
	Gateway   [4]byte
	DHCPState byte
	LeaseIP   [4]byte
	LeaseMask [4]byte
	// LeaseServer / LeaseSecs
	LeaseServer [4]byte
	LeaseSecs   uint32
	// TCP
	TCPState    byte
	TCPPeerIP   [4]byte
	TCPPeerPort uint16
	// UDP
	UDPCount byte
	UDPPorts [4]uint16
	// ARP
	ARPCount byte
	ARPIPs   [4][4]byte
	ARPMACs  [4][6]byte
	// Counters
	TXFrames   uint64
	TXBytes    uint64
	RXFrames   uint64
	RXBytes    uint64
	RXFiltered uint64
	RXOverflow uint64
	// TCP segment counters: syn_sent, synack_recv, ack_sent, data_sent,
	// data_recv, fin_sent, finack_recv, rst_sent.
	TCPSegs [8]uint64
	// UDP datagram counters: received, sent, loopbacked, dropped.
	UDPDgrams [4]uint64
}

// netStatsOffsets are the layout pins, byte-for-byte the set the kernel
// comment and user/src/lib/netstats.zig pin (the Go compiler inserts the same
// natural-alignment padding as the Zig extern struct, which is what makes a
// field-by-field Go mirror correct rather than merely plausible).
type netStatsOffsets struct {
	OwnIP, Gateway, DHCPState, LeaseSecs uintptr
	TCPState, TCPPeerIP, TCPPeerPort     uintptr
	UDPCount, TXFrames, RXBytes          uintptr
	TCPSegs, UDPDgrams                   uintptr
	Size                                 uintptr
}

func netStatsLayout() netStatsOffsets {
	var s NetStats
	return netStatsOffsets{
		OwnIP:       unsafe.Offsetof(s.OwnIP),
		Gateway:     unsafe.Offsetof(s.Gateway),
		DHCPState:   unsafe.Offsetof(s.DHCPState),
		LeaseSecs:   unsafe.Offsetof(s.LeaseSecs),
		TCPState:    unsafe.Offsetof(s.TCPState),
		TCPPeerIP:   unsafe.Offsetof(s.TCPPeerIP),
		TCPPeerPort: unsafe.Offsetof(s.TCPPeerPort),
		UDPCount:    unsafe.Offsetof(s.UDPCount),
		TXFrames:    unsafe.Offsetof(s.TXFrames),
		RXBytes:     unsafe.Offsetof(s.RXBytes),
		TCPSegs:     unsafe.Offsetof(s.TCPSegs),
		UDPDgrams:   unsafe.Offsetof(s.UDPDgrams),
		Size:        unsafe.Sizeof(s),
	}
}

// netStatsWanted is the offset table the kernel + Zig mirror agree on.
func netStatsWanted() netStatsOffsets {
	return netStatsOffsets{
		OwnIP: 6, Gateway: 10, DHCPState: 14, LeaseSecs: 28,
		TCPState: 32, TCPPeerIP: 33, TCPPeerPort: 38, UDPCount: 40,
		TXFrames: 96, RXBytes: 120, TCPSegs: 144, UDPDgrams: 208,
		Size: 240,
	}
}

// NetStatsSnapshot fetches one live snapshot (slot 62). ok is false when the
// kernel refused (a short buffer / off the guest), in which case the caller
// keeps its previous view — the same contract Zig's read_stats has.
func NetStatsSnapshot() (NetStats, bool) {
	var s NetStats
	buf := unsafe.Slice((*byte)(unsafe.Pointer(&s)), NetStatsBytes)
	r := svc2(SlotNetStats, uintptr(unsafe.Pointer(&buf[0])), uintptr(len(buf)))
	if r != NetStatsBytes {
		return NetStats{}, false
	}
	return decodeNetStats(buf), true
}

// decodeNetStats unpacks a wire snapshot. It is the single decode site, so a
// host test can drive it with a synthesised buffer (the guest only ever hands
// it the kernel's own bytes).
func decodeNetStats(b []byte) NetStats {
	if len(b) < NetStatsBytes {
		return NetStats{}
	}
	var s NetStats
	copy(s.MAC[:], b[0:6])
	copy(s.OwnIP[:], b[6:10])
	copy(s.Gateway[:], b[10:14])
	s.DHCPState = b[14]
	copy(s.LeaseIP[:], b[15:19])
	copy(s.LeaseMask[:], b[20:24])
	copy(s.LeaseServer[:], b[24:28])
	s.LeaseSecs = getU32(b[28:])
	s.TCPState = b[32]
	copy(s.TCPPeerIP[:], b[33:37])
	s.TCPPeerPort = getU16(b[38:])
	s.UDPCount = b[40]
	for i := 0; i < 4; i++ {
		s.UDPPorts[i] = getU16(b[42+2*i:])
	}
	s.ARPCount = b[50]
	for i := 0; i < 4; i++ {
		copy(s.ARPIPs[i][:], b[51+4*i:55+4*i])
	}
	for i := 0; i < 4; i++ {
		copy(s.ARPMACs[i][:], b[67+6*i:73+6*i])
	}
	s.TXFrames = getU64(b[96:])
	s.TXBytes = getU64(b[104:])
	s.RXFrames = getU64(b[112:])
	s.RXBytes = getU64(b[120:])
	s.RXFiltered = getU64(b[128:])
	s.RXOverflow = getU64(b[136:])
	for i := 0; i < 8; i++ {
		s.TCPSegs[i] = getU64(b[144+8*i:])
	}
	for i := 0; i < 4; i++ {
		s.UDPDgrams[i] = getU64(b[208+8*i:])
	}
	return s
}
